### ISSUE TITLE
Where should I place express-session in the chain?

### ISSUE THREAD

**Issue by mercmobily** [Polymer/prpl-server#66]
Hi,

I am starting a new project, and I am using express-session for the first time.
At the moment, I am doing app.use() like so:

    // PART ONE: SET THE SESSION

    // Make up the session
    app.use(session({
      key: 'proj',
      secret: '(deleted)',
      store: sessionStore,
      resave: false,
      saveUninitialized: false
    }))

    // PART 2: SET SOME API ENTRY POINTS

    // This WILL set routes. Assume: /cargos
    var cargos = require('./stores/Cargos.js')
    cargos.protocolListenHTTP({app: app})

    PART 3: PRPL-SERVER

    app.get('/*', prpl.makeHandler('./public/build', {
      builds: [
        { name: 'es6-unbundled', browserCapabilities: ['es2015', 'push'] },
        { name: 'es6-bundled', browserCapabilities: ['es2015'] },
        { name: 'es5-bundled' }
      ]
    }))
  }

The problem with this approach is that the session is created for _each_ served file. In a conventional server, I would put the middleware for the session _after_ the serving of static files. However, here node-prpl-server is dealing with static files -- and is in fact _always_ serving something, effectively.

Now:

* If I put `app.use(session({` _after_  `app.get('/*', prpl.makeHandler(`, my `cargo` routes won't get the session information (which is necessary)

* If I put `app.use(session({` _after_  `app.get('/*', prpl.makeHandler(` at the beginning, before session and cargo routes, the module will _always_ serve anything and everything.

A possible "solution" I worked out is this:

    // PART 1: PRPL-SERVER *EXCEPT* /stores

    app.get('/*', (req, res, done) => 
      if (req.url.startsWith('/stores/')) return next(null)
      prpl.makeHandler('./public/build', {
      builds: [
        { name: 'es6-unbundled', browserCapabilities: ['es2015', 'push'] },
        { name: 'es6-bundled', browserCapabilities: ['es2015'] },
        { name: 'es5-bundled' }
      ]
    }))

    // PART 2: SET THE SESSION

    // Make up the session
    app.use(session({
      key: 'proj',
      secret: '(deleted)',
      store: sessionStore,
      resave: false,
      saveUninitialized: false
    }))

    // PART 3: SET SOME API ENTRY POINTS

    // This WILL set routes. Assume: /stores/cargos
    var cargos = require('./stores/Cargos.js')
    cargos.protocolListenHTTP({app: app})

Basically, I am leaving `/stores/*` out of the node-prpl equation.

Questions:

1) Is this a sane way to go about it?
1a) I *could* probably place the conditional around `app.use(session({`

2)
Right now, every time a user logs in, the app greys out until an ajax call is completed and the user's config is retrieved. This is not ideal: it means that the user is greeted with a "loading" message (while the app retrieves the user config) plus all of the error management relating to this.

I was thinking of serialising the user's configuration within the served `index` page served by node-prpl. Is that advisable and indeed feasible? OR should index.html _never_ change?

**Comment by mercmobily** [comment]
Silly me, I created an endpoint which returns an ajax statement `window.userConfig = { ... }` (which has info like userId, loggedIn flag, etc.) and made sure that index.html loads it. So there: the index doesn't change, but the app knows things right away.

**Comment by mercmobily** [comment]
I asked this more than 1 year ago... I did end up doing this: `if (req.url.startsWith('/stores/')) return next(null)` but it feels terrible. Plus, `/stores/` is not even the only one that gets filtered...

What's a sane way to deal with this?

**Comment by mercmobily** [comment]
I cannot reproduce this problem with the latest session. Closing for now.
